### PR TITLE
Surrogates (fixes #400)

### LINKED ISSUE #400: Script surrogates?
It seems that NoScript and Ghostery have worked around a lot of script-blocking breakage by writing drop in replacement "surrogate" scripts for common third parties.  The NoScript version is open source, and mirrored on GitHub [here](https://github.com/avian2/noscript/blob/master/xpi/defaults/preferences/noscript.js).

We should consider whether we can transpose some or all of this set of scripts straight into socialwidgets.json (or a separate semantically equivalent file); it might save us a lot of breakage with the extra non-iframed script blocking we're going to be doing, and also allow us to shrink the yellowlist.

### PR DESCRIPTION
**Pull request by ghostwords**
WIP pull request for #400 

Besides review, what's left is more surrogates and more testing. We should run some queries against the "report broken site" db to get some recent high-profile breakages we can surrogate.

### CODE REVIEW

File: data/surrogates.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Privacy Badger.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+require.scopes.surrogatedb = (function() {
+
+const hostnames = {
+  'www.google-analytics.com': [
```
Comment (cooperq) [inline data/surrogates.js]: This data structure should be better commented.

File: data/surrogates.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    '/ga.js',
+  ],
+};
+
+const surrogates = {
```
Comment (cooperq) [inline data/surrogates.js]: More documentation needed here as well.

File: src/surrogates.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+require.scopes.surrogates = (function() {
+
+const db = require('surrogatedb');
+
+// gets called within request-blocking listeners
+// needs to be fast!
+function getSurrogateURI(script_url, script_hostname) {
+  // do we have an entry for the script hostname?
+  if (db.hostnames.hasOwnProperty(script_hostname)) {
+    const hosts = db.hostnames[script_hostname];
```
Comment (cooperq) [inline src/surrogates.js]: needs more doc

Comment (cooperq) [inline src/surrogates.js]: rename to tokens

File: tests/tests/utils.js [hunk 4]
```diff
@@ -90,4 +91,9 @@
     ok(mdfp.isMultiDomainFirstParty("google.com", "youtube.com"));
     ok(!mdfp.isMultiDomainFirstParty("google.com", "nyt.com"));
   });
+
+  test("surrogate script URL lookups", function() {
+    // URLs that should have a surrogate
+    ok(!!getSurrogateURI('http://www.google-analytics.com/ga.js', 'www.google-analytics.com'));
```
Comment (cooperq) [inline tests/tests/utils.js]: Test for no surrogate?
Test other  surrogates...

Comment (cooperq) [inline tests/tests/utils.js]: test that query strings are handled in a sane way

Comment (cooperq) [inline tests/tests/utils.js]: selenium integration test

**Comment by ghostwords** [comment]
@cooperq OK, should be good to merge now. Take a look please.
Will need to later:
- Update the integration test URL from `eff.org` to `pbtest.org` at some point.
- Replace `pb` by `badger` in the integration test after #951 is merged. Depends on which pull request is merged into master first I guess.

**Comment by ghostwords** [comment]
@gunesacar Let me know if you have any suggestions to improve the [script surrogates integration test](https://github.com/EFForg/privacybadgerchrome/blob/f11cdf6996a222d04cc15852fb5e06ed155db605/tests/selenium/surrogates_test.py).

**Comment by cooperq** [comment]
Looks good to me!

**Comment by gunesacar** [comment]
> @gunesacar Let me know if you have any suggestions to improve the script surrogates integration test.

It looks pretty good to me.

Maybe you could consider adding a test for outbound link tracking:
https://support.google.com/analytics/answer/1136920?hl=en

**Pull request merged**